### PR TITLE
[IMP] *: ModelFieldSelector improvements

### PR DESCRIPTION
**Pull request by Polymorphe57**
We refactor the ModelFieldSelector and ModelFieldSelectorPopover
components.

We also improve a bit ModelFieldSelectorPopover:
    - on first page, the button to go back is no longer available
      (so that it is now more difficult to produce an invalid path)
    - we always start with a page presenting the model where the last
      selected field name belongs to
    - the keyboard navigation is improved
    - click on model field selector opens the popover with the focus in
      the search input (if any)
    - for relational fields in popover: the user can either click on the
      relational field (and select it) or a special button that make him
      follow the relation to the field comodel

We also refactor the hook useDynamicPlaceholder to make it use a new
component DynamicPlaceholderPopover that uses ModelFieldSelectorPopover.

Task ID: 3272798

Related Enterprise PR: https://github.com/odoo/enterprise/pull/39673

Co-authored-by: Mathieu Duckerts-Antoine <dam@odoo.com>
Co-authored-by: Michaël Mattiello <mcm@odoo.com>

### CODE REVIEW

File: addons/web/static/src/core/model_field_selector/model_field_selector.js [hunk 1]
```diff
@@ -1,99 +1,120 @@
...
+    async _updatePath(resModel, path) {
+        const state = await this.loadPath(resModel, path);
+        Object.assign(this.state, state);
     }
 
-    getFieldNameChain(fieldName) {
-        return fieldName.length ? fieldName.split(".") : [];
+    async updatePath(resModel, path) {
+        const state = await this.keepLast.add(this.loadPath(resModel, path));
+        Object.assign(this.state, state);
```
Comment (mcm-odoo) [inline addons/web/static/src/core/model_field_selector/model_field_selector.js:81]: we can factorize that
```js
async updatePath(resModel, path, isConcurrent) {
  let prom = this.loadPath(resModel, path);
  if (isConcurrent) {
    prom = this.keepLast.add(prom);
  }
  const state = await prom;
  Object.assign(this.state, state);
}
```

File: addons/web/static/src/core/domain_selector/domain_selector.js [hunk 2]
```diff
@@ -154,13 +149,18 @@ export class DomainSelector extends Component {
...
+    }
+
     updateBranchOperator(node, operator) {
         node.operator = operator;
         this.notifyChanges();
     }
 
-    async updateField(node, field) {
-        const fieldDef = await this.loadFieldDef(this.props.resModel, field);
+    async updateField(node, field, { fieldDef }) {
```
Comment (mcm-odoo) [inline addons/web/static/src/core/domain_selector/domain_selector.js:161]: remove `async`

File: addons/web/static/src/core/domain_selector/domain_selector.js [hunk 3]
```diff
@@ -154,13 +149,18 @@ export class DomainSelector extends Component {
...
+        fieldDef ||= { type: "integer" };
```
Comment (mcm-odoo) [inline addons/web/static/src/core/domain_selector/domain_selector.js:162]: first time I see someone use this operator

File: addons/web/static/src/core/domain_selector/domain_selector.js [hunk 4]
```diff
@@ -154,13 +149,18 @@ export class DomainSelector extends Component {
...
+        field = fieldDef ? field : "";
```
Comment (mcm-odoo) [inline addons/web/static/src/core/domain_selector/domain_selector.js:163]: will always be truthy because of the previous line
Am I wrong?

Comment (Polymorphe57) [reply addons/web/static/src/core/domain_selector/domain_selector.js:163]: no, you're right :-)

File: addons/web/static/src/core/domain_selector/domain_selector.js (hunk 2 above)
Comment (Polymorphe57) [reply addons/web/static/src/core/domain_selector/domain_selector.js:161]: done

File: addons/web/static/src/core/model_field_selector/model_field_selector.js (hunk 1 above)
Comment (Polymorphe57) [reply addons/web/static/src/core/model_field_selector/model_field_selector.js:81]: done

File: addons/web/static/src/core/popover/popover_hook.js
[the remark this answers is not in the document]
Comment (Polymorphe57) [reply addons/web/static/src/core/popover/popover_hook.js:1]: done

File: addons/web/static/src/core/domain_selector/domain_selector.js (hunk 4 above)
Comment (Polymorphe57) [reply addons/web/static/src/core/domain_selector/domain_selector.js:163]: order changed

File: addons/web/static/src/core/model_field_selector/model_field_selector.js [hunk 5]
```diff
@@ -1,99 +1,119 @@
...
             return;
         }
-        this.popover.open(ev.currentTarget, {
-            chain: this.state.chain,
-            update: this.update.bind(this),
+        this.newPath = null;
+        this.popover.open(currentTarget, {
+            resModel: this.props.resModel,
+            path: this.props.path,
+            update: async (path) => {
```
Comment (ged-odoo) [inline addons/web/static/src/core/model_field_selector/model_field_selector.js:109]: why async?

File: addons/web/static/src/core/model_field_selector/model_field_selector_popover.js [hunk 6]
```diff
@@ -1,202 +1,273 @@
...
+        this.state.page = page;
+        this.state.page.searchFields();
+        this.props.update(page.path);
+    }
+
+    searchFields(query) {
+        this.state.page.searchFields(query);
+    }
+
+    async selectField(field) {
```
Comment (ged-odoo) [inline addons/web/static/src/core/model_field_selector/model_field_selector_popover.js:202]: why async

File: addons/web/static/src/core/model_field_selector/model_field_selector_popover.js [hunk 7]
```diff
@@ -1,202 +1,273 @@
...
+        this.keepLast.add(Promise.resolve());
+        this.state.page.selectedName = field.name;
+        this.props.update(this.state.page.path);
+        this.props.close();
+    }
+
+    async onDebugInputKeydown(ev) {
```
Comment (ged-odoo) [inline addons/web/static/src/core/model_field_selector/model_field_selector_popover.js:209]: why async?

File: addons/web/static/src/core/model_field_selector/model_field_selector_popover.js [hunk 8]
```diff
@@ -1,202 +1,276 @@
...
+        this.previousPage = previousPage;
+        this.selectedName = selectedName;
+
+        this.sortedFieldNames = sortBy(Object.keys(fieldDefs), (key) => fieldDefs[key].string);
+
+        this.fieldNames = this.sortedFieldNames;
+        this.query = "";
+        this.focusedFieldName = null;
+        this.searchFields();
+    }
```
Comment (mcm-odoo) [inline addons/web/static/src/core/model_field_selector/model_field_selector_popover.js:23]:
```suggestion
        this.resModel = resModel;
        this.fieldDefs = fieldDefs;
        this.previousPage = previousPage;
        this.selectedName = selectedName;
        this.sortedFieldNames = sortBy(Object.keys(fieldDefs), (key) => fieldDefs[key].string);
        this.fieldNames = this.sortedFieldNames;
        this.query = "";
        this.focusedFieldName = null;
        this.resetFocusedFieldName();
    }
```

File: addons/web/static/src/views/fields/dynamic_placeholder_hook.js [hunk 9]
```diff
@@ -1,36 +1,34 @@
...
     const ownerField = useComponent();
     const triggerKeyReplaceRegex = new RegExp(`${TRIGGER_KEY}$`);
     let closeCallback;
     let positionCallback;
-    const popover = usePopover(ModelFieldSelectorPopover, {
+    const popover = usePopover(DynamicPlaceholderPopover, {
+        closeOnClickAway: true,
         onclose: () => closeCallback?.(),
         onPositioned: () => positionCallback?.(),
     });
```
Comment (mcm-odoo) [inline addons/web/static/src/views/fields/dynamic_placeholder_hook.js:18]:
```suggestion
    const popover = usePopover(DynamicPlaceholderPopover, {
        onclose: () => closeCallback?.(),
        onPositioned: () => positionCallback?.(),
    });
```

**Comment by aab-odoo** [comment]
robodoo delegate=mcm-odoo

**Comment by mcm-odoo** [comment]
@robodoo rebase-ff r+

**Pull request closed without merging**